workflow/artifacts/oss/oss.go

```diff
@@ -148,6 +152,53 @@ func (ossDriver *ArtifactDriver) ListObjects(artifact *wfv1.Artifact) ([]string,
 	return files, err
 }
 
+func (ossDriver *ArtifactDriver) setBucketLifecycleRule(client *oss.Client, artifact *wfv1.Artifact) error {
+	if artifact.OSS.LifecycleRule.MarkInfrequentAccessAfterDays == nil && artifact.OSS.LifecycleRule.MarkDeletionAfterDays == nil {
+		return nil
+	}
+	var markInfrequentAccessAfterDays int
+	var markDeletionAfterDays int
+	if artifact.OSS.LifecycleRule.MarkInfrequentAccessAfterDays != nil {
+		markInfrequentAccessAfterDays = int(*artifact.OSS.LifecycleRule.MarkInfrequentAccessAfterDays)
+	}
+	if artifact.OSS.LifecycleRule.MarkDeletionAfterDays != nil {
+		markDeletionAfterDays = int(*artifact.OSS.LifecycleRule.MarkDeletionAfterDays)
+	}
+	if markInfrequentAccessAfterDays > markDeletionAfterDays {
+		return fmt.Errorf("markInfrequentAccessAfterDays cannot be large than markDeletionAfterDays")
+	}
+
+	// Set expiration rule.
+	expirationRule := oss.BuildLifecycleRuleByDays("expiration-rule", artifact.OSS.Key, true, markInfrequentAccessAfterDays)
+	// Automatically delete the expired delete tag so we don't have to manage it ourselves.
+	deleteMark := true
+	expiration := oss.LifecycleExpiration{
+		ExpiredObjectDeleteMarker: &deleteMark,
+	}
+	// Convert to Infrequent Access (IA) storage type for objects that are expired after a period of time.
+	versionTransition := oss.LifecycleVersionTransition{
+		NoncurrentDays: markInfrequentAccessAfterDays,
+		StorageClass:   oss.StorageIA,
+	}
+	// Mark deletion after a period of time.
+	versionExpiration := oss.LifecycleVersionExpiration{
+		NoncurrentDays: markDeletionAfterDays,
+	}
+	versionTransitionRule := oss.LifecycleRule{
```

Comment:
might it be good idea to have a 1-2-1 structural mapping between `wfv1.LifecycleRule` and `oss.LifecycleRule`? that way you could have copy the fields

Reply:
It's intentional so it's easier for users to just specify the number of days. This already supports the most popular use cases to save cloud storage expenses for users.

Reply:
can it be extended to more complex cases easily?

Reply:
It might be unnecessary to support more complex cases at this point. This should cover most of the common use cases when handling 0SS objects created by Argo. Other complicated usages such as tagging and matching on objects should be done using OSS CLI or SDK (we could introduce new fields in `OSSLifecycleRule` to support those as well when needed).